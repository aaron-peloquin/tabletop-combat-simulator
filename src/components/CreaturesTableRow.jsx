import {Button, TableRow, TableCell} from "@material-ui/core"
import {Edit, Delete} from "@material-ui/icons"
import Link from "next/link"
import {connect} from "react-redux"

import deleteCreature from "./../store/action/DeleteCreature"

const CreaturesTableRow = (props) => {
  const {creature, funcDeleteCreature, dispatch} = props
  return <TableRow key={creature.hash}>
    <TableCell>
      <Link as={`/creature/${creature.hash}`} href={`/creature?hash=${creature.hash}`}>
        <a>
          <Button variant="contained" color="primary">
            <Edit />
          </Button>
        </a>
      </Link>
      <Button variant="contained" color="secondary" onClick={()=>{
        funcDeleteCreature(dispatch, creature.hash)
      }}>
        <Delete />
      </Button>
    </TableCell>
    <TableCell>{creature.name}</TableCell>
    <TableCell>{creature.cr}</TableCell>
  </TableRow>
}

CreaturesTableRow.defaultProps = {
  funcDeleteCreature: deleteCreature,
}

export default connect()(CreaturesTableRow)
