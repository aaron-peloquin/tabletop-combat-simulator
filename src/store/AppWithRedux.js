import React from 'react'
import {initializeStore} from './store'

const isServer = typeof window === 'undefined'
const reduxStoreSymbol = Symbol("reduxStore")

function getOrCreateStore (initialState) {
  // Always make a new store if server, otherwise state is shared between requests
  if (isServer) {
    return initializeStore(initialState)
  }

  // Create store if unavailable on the client and set it on the window object
  if (!window[reduxStoreSymbol]) {
    window[reduxStoreSymbol] = initializeStore(initialState)
  }
  return window[reduxStoreSymbol]
}

export default (App) => {
  return class AppWithRedux extends React.Component {
    static async getInitialProps (appContext) {
      // Get or Create the store with `undefined` as initialState
      // This allows you to set a custom default initialState
      const reduxStore = getOrCreateStore()

      // Provide the store to getInitialProps of pages
      appContext.ctx.reduxStore = reduxStore

      let appProps = {}
      if (typeof App.getInitialProps === 'function') {
        appProps = await App.getInitialProps(appContext)
      }

      return {
        ...appProps,
        initialReduxState: reduxStore.getState()
      }
    }

    constructor (props) {
      super(props)
      this.reduxStore = getOrCreateStore(props.initialReduxState)
    }

    render () {
      return <App {...this.props} reduxStore={this.reduxStore} />
    }
  }
}
