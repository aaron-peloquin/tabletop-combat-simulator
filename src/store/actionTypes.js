/**
 * This is an object of Redux Actions. It mostly exists to
 *  ensure we do not create duplicate actions, but it can
 *  also be used as a list of our existing actions
*/
const actions = {
  CopyCreature: 'CopyCreature',
  DeleteAllCreatures: 'DeleteAllCreatures',
  DeleteCreature: 'DeleteCreature',
  SaveCreature: 'SaveCreature',
  SetEditCreature: 'SetEditCreature',
  ToggleSidebar: 'ToggleSidebar',
  RunSimulation: 'RunSimulation',
  UpdateEditCreature: 'UpdateEditCreature',
}

export default actions
