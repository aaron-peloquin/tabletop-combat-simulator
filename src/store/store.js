import {createStore, applyMiddleware, combineReducers} from "redux"
import {composeWithDevTools} from "redux-devtools-extension"
import thunkMiddleware from "redux-thunk"

/** Reducers */
import combat from "./reducer/combat"
import creatures from "./reducer/creatures"
import sideBar from "./reducer/sideBar"
import editing from "./reducer/editing"

/** Reducers */
const reducers = combineReducers({
  combat,
  creatures,
  editing,
  sideBar,
})

/**
 * Initializes the store
 * @param {obj} initialState the state our app starts with
 * @return {obj} the redux store
 */
export function initializeStore(initialState = initialState) {
  return createStore(reducers, initialState, composeWithDevTools(applyMiddleware(thunkMiddleware)))
}
