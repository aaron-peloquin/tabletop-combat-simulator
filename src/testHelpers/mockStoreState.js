export default {
  "creatures": [
    {
      "name": "Goplin",
      "cr": "1/4",
      "hash": "hivn3p0k",
      "description": "A stout monster",
      "defaultInitiative": 10,
      "armor": 11,
      "hp": 5,
      "hitDiceEquation": "1d20+1",
      "damageDiceEquation": "1d4"
    },
    {
      "name": "Elven Ranger",
      "cr": "1/2",
      "hash": "h3dxf5wl",
      "description": "A swift hunter",
      "defaultInitiative": 16,
      "armor": 13,
      "hp": 10,
      "hitDiceEquation": "1d20+8",
      "damageDiceEquation": "1d6+3"
    },
    {
      "name": "Knight",
      "cr": "1",
      "hash": "hee9p286",
      "description": "An honorable soldier",
      "defaultInitiative": 11,
      "armor": 18,
      "hp": 25,
      "hitDiceEquation": "1d20+4",
      "damageDiceEquation": "1d12+4"
    },
    {
      "name": "Lizardfolk",
      "cr": "3",
      "hash": "howx26rt",
      "description": "An ancient sneaky monster",
      "defaultInitiative": 14,
      "armor": 14,
      "hp": 15,
      "hitDiceEquation": "1d20+5",
      "damageDiceEquation": "1d10+2"
    }
  ],
  "sideBar": false,
}