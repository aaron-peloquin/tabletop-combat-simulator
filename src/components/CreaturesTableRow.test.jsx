import {Provider} from "react-redux"
import {createShallow} from "@material-ui/core/test-utils"
import toJson from "enzyme-to-json"
import {initializeStore} from "../store/store"

import CreaturesTableRow from "./CreaturesTableRow"
import mockStoreState from "../testHelpers/mockStoreState"

describe("<CreaturesTableRow />", ()=>{
  const renderShallowUntilComponent = createShallow({"untilSelector": "CreaturesTableRow"})
  let store; let props; let Component
  beforeEach(async ()=>{
    store = initializeStore(mockStoreState)
    store.dispatch = jest.fn(store.dispatch)
    props = {creature: mockStoreState.creatures[0]}
    Component = renderShallowUntilComponent(<Provider store={store}><CreaturesTableRow {...props} /></Provider>)
  })

  it("loads", () => {
    expect(typeof Component).toBe("object")
  })

  it("has a link back to the creatures page", () => {
    expect(Component.find(`Link[href="/creature?hash=${props.creature.hash}"] a`).length).toBeGreaterThanOrEqual(1)
  })

  it("has an edit icon inside the link", () => {
    expect(Component.find(`Link[href="/creature?hash=${props.creature.hash}"] a pure(EditIcon)`).length).toBe(1)
  })

  it("has a delete icon and button", () => {
    expect(Component.find("WithStyles(Button) pure(DeleteIcon)").length).toBe(1)
  })

  it("calls dispatch on click of delete", () => {
    Component.find("WithStyles(Button) pure(DeleteIcon)").parent().simulate("click")
    expect(store.dispatch).toBeCalledTimes(1)
  })

  it("snapshots", () => {
    expect(toJson(Component)).toMatchSnapshot()
  })
})
