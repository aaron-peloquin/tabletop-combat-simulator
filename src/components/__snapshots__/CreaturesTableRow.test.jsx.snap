// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreaturesTableRow /> snapshots 1`] = `
<WithStyles(TableRow)
  key="hivn3p0k"
>
  <WithStyles(TableCell)>
    <Link
      as="/creature/hivn3p0k"
      href="/creature?hash=hivn3p0k"
    >
      <a>
        <WithStyles(Button)
          color="primary"
          variant="contained"
        >
          <pure(EditIcon) />
        </WithStyles(Button)>
      </a>
    </Link>
    <WithStyles(Button)
      color="secondary"
      onClick={[Function]}
      variant="contained"
    >
      <pure(DeleteIcon) />
    </WithStyles(Button)>
  </WithStyles(TableCell)>
  <WithStyles(TableCell)>
    Goplin
  </WithStyles(TableCell)>
  <WithStyles(TableCell)>
    1/4
  </WithStyles(TableCell)>
</WithStyles(TableRow)>
`;
